### PR TITLE
fix int – float equality for big ints, inf and nan

### LINKED ISSUE #373: int with float equality should not compare only floats
Right now, when you check equality of an int with a float, both are cast to floats, and then the equality check happens on them. This is problematic because bigger ints might not be represented exactly as 64 floats (and thus many ints might look equal to a given float). Eg. `1e308` is a valid float which does **not** have an exact 10³⁸ value, in both CPython and RustPython:

```
>>>>> int(1e308)
100000000000000001097906362944045541740492309677311846336810682903157585404911491537163328978494688899061249669721172515611590283743140088328307009198146046031271664502933027185697489699588559043338384466165001178426897626212945177628091195786707458122783970171784415105291802893207873272974885715430223118336
```

but when you try to compare `10**308` with it, RustPython does not notice that:
```
>>>>> 10**308 == 1e308
True
>>>>> # and other ints…
>>>>> 10**308 + 30 == 1e308
True
```

while CPython handles this correctly:
```
>>> 10**308 == 1e308
False
>>> 10**308 + 30 == 1e308
False
>>> 100000000000000001097906362944045541740492309677311846336810682903157585404911491537163328978494688899061249669721172515611590283743140088328307009198146046031271664502933027185697489699588559043338384466165001178426897626212945177628091195786707458122783970171784415105291802893207873272974885715430223118336 == 1e308
True
```

I am not sure how CPython handles it and what is the best approach, but perhaps doing two checks, on both casts (basically as if `int1 == float2` were desugarized to `int1 == int(float2) and float(int1) == float2`) would be an OK solution?

### PR DESCRIPTION
**Pull request by silmeth**
Fixes #373

### CODE REVIEW

**Comment by silmeth** [comment]
I don’t understand why the build failed. The job’s log says that it failed tests on CPython with:

```
Traceback (most recent call last):
  File "/home/travis/build/RustPython/RustPython/tests/snippets/int_float_equality.py", line 12, in <module>
    assert not (10**500 == math.inf)
AttributeError: module 'math' has no attribute 'inf'
```

using Python 3.6, but the Python docs say that [`math.inf`](https://docs.python.org/3/library/math.html#math.inf) is available since Python 3.5, and it works on my machine with Python 3.7…

Should I change `math.inf` in the test to `float('inf')`?

**Comment by cthulahoops** [comment]
There's a math.py in the snippets directory. Is that getting imported in preference to the math library?

**Comment by silmeth** [comment]
It might be that… but then, why does it work on my machine? I run the tests in the same repo, basically the same way (`cd tests` and `pipenv run pytest -v`). Could it be that different Python versions/builds have different precedence for importing modules?

I might try to rename the `math.py` test suite to fix it, but I’d like first reproduce the problem locally to be sure it works.

**Comment by silmeth** [comment]
OK, no idea how to reproduce this, let’s try removing `math.py`…

**Comment by windelbouwman** [comment]
I think you hit an import mechanism bug. When importing the math module, it will pick the math.py file in the snippets directory. We should rename this snippet, so that import math less ambiguous. I think the wrong math module got imported (the test script), and that one does not have an `inf` member.

File: tests/snippets/math.py [hunk 1]
```diff
@@ -1,29 +0,0 @@
-a = 4
```
Comment (windelbouwman) [inline tests/snippets/math.py]: Instead of deletion, please rename this test snippet.

Comment (silmeth) [reply tests/snippets/math.py]: What would be a good name? `math_ops_and_funcs`, `mathematics`, `maths`, `math_basics`?

**Comment by silmeth** [comment]
> I think the wrong math module got imported (the test script), and that one does not have an `inf` member.

I see that now, but I still don’t understand why I don’t hit the same problem when locally running the tests with CPython 3.6.7 on my (Kubuntu 18.10) PC. For some reason here the standard math library is imported…

File: tests/snippets/math.py (hunk 1 above)
Comment (silmeth) [reply tests/snippets/math.py]: Also, should I amend my last commit or just make another one re-adding the file?

Comment (windelbouwman) [reply tests/snippets/math.py]: If you want you can amend your commit, or create a new one. I do not have a strong preference.

**Comment by windelbouwman** [comment]
I think the import mechanism is different between rustpython and cpython. cpython gives precedence to builtin modules it appears.

**Comment by silmeth** [comment]
> I think the import mechanism is different between rustpython and cpython. cpython gives precedence to builtin modules it appears.

But I am writing about different behaviour of the CPython only. It failed on the Travis worker, while all tests with CPython work on my machine. They should fail on my end too (then I would catch this earlier and rename the file before sending a PR), but they didn’t. My CPython happily passed the `int_float_equality` test case…

**Comment by windelbouwman** [comment]
This indeed is strange. Maybe it is a cpython bug?

**Pull request merged**